src/btf_parser.cpp

```diff
@@ -1081,3 +1085,393 @@ std::map<std::string, size_t> parse_btf_map_sections(const btf_type_data& btf_da
     }
     return map_offset_to_descriptor_index;
 }
+
+/**
+ * @brief Write a type into a vector and grow it.
+ *
+ * @tparam T Type to write
+ * @param[in] btf The .BTF section
+ * @param[in] value The value to write
+ */
+template <typename T>
+void _write_btf(std::vector<std::byte>& btf, const T& value) {
+    size_t length = 0;
+    size_t offset = btf.size();
+    if constexpr (std::is_same<T, std::string>::value) {
+        length = value.length();
+        btf.resize(offset + length + 1);
+        memcpy(btf.data() + offset, value.c_str(), length + 1);
+    } else if constexpr (std::is_same<T, std::vector<std::byte>>::value) {
+        length = value.size();
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, value.data(), length);
+    } else {
+        length = sizeof(T);
+        btf.resize(offset + length);
+        memcpy(btf.data() + offset, &value, length);
+    }
+}
+
+std::vector<std::byte> btf_serialize_types(const std::vector<btf_kind>& btf_kind) {
+    std::vector<std::byte> btf;
+    std::vector<std::byte> string_table_bytes;
+    std::map<std::string, uint32_t> string_table_map;
+    std::vector<std::byte> type_table_bytes;
+
+    auto string_to_offset = [&](const std::optional<std::string>& str) -> uint32_t {
+        if (!str) {
+            return 0;
+        }
+        auto it = string_table_map.find(*str);
+        if (it != string_table_map.end()) {
+            return it->second;
+        }
+        size_t offset = string_table_bytes.size();
+        _write_btf(string_table_bytes, *str);
+        string_table_map[*str] = static_cast<uint32_t>(offset);
+        return static_cast<uint32_t>(offset);
+    };
```

Comment:
It wasn't obvious to me where else are string_table_map and string_table_bytes accessed. I suggest using a dedicated class (though there are probably better names):
```cpp
class StringCache {
    std::map<std::string, uint32_t> string_table_map{};
    std::vector<std::byte> string_table_bytes{};
  public:

    StringCache() { lookup(""); };

    uint32_t lookup(const std::optional<std::string>& str) {
        if (!str) {
            return 0;
        }
        uint32_t offset = static_cast<uint32_t>(string_table_bytes.size());
        auto [it, did_not_exist] = string_table_map.try_emplace(*str, offset);
        if (did_not_exist) {
            write_btf(string_table_bytes, *str);
        }
        return it->second;
    }
    
    const std::vector<std::byte>& get_table_bytes() {
        return string_table_bytes;
    }
};
```

(We have a similar cache/hash used for variable_t).